apm-agent-plugins/apm-vertx/apm-vertx-4-plugin/src/main/java/co/elastic/apm/agent/vertx_4/VertxWebInstrumentation.java

```diff
@@ -0,0 +1,146 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.vertx_4;
+
+import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.sdk.advice.AssignTo;
+import co.elastic.apm.agent.sdk.state.CallDepth;
+import co.elastic.apm.agent.vertx_4.tracer.ElasticVertxTracer;
+import co.elastic.apm.agent.vertx_4.tracer.ElasticVertxTracerWrapper;
+import co.elastic.apm.agent.vertx_4.wrapper.RoutingContextHandlerWrapper;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.spi.tracing.VertxTracer;
+import io.vertx.ext.web.RoutingContext;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.isPrivate;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.namedOneOf;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public abstract class VertxWebInstrumentation extends TracerAwareInstrumentation {
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        // ensure only Vertx versions >= 4.0 are instrumented
+        return classLoaderCanLoadClass("io.vertx.core.http.impl.Http1xServerRequest");
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Arrays.asList("vertx", "vertx-web");
+    }
+
+    /**
+     * Instruments {@link io.vertx.core.impl.VertxFactory#createTracer}}
+     * <p>
+     * Instruments VertxFactory to return our own Tracer implementation.
+     */
+    public static class VertxFactoryInstrumentation extends VertxWebInstrumentation {
```

Comment:
It looks like this instrumentation depends on being executed while Vert.x initializes and thus doesn't work with runtime attachment.
I have an idea for an alternative, though. You could instrument `io.vertx.core.impl.ContextInternal#tracer` and return a constant noop tracer if the application didn't register a tracer. Then, instrument implementations of `io.vertx.core.spi.tracing.VertxTracer`. Instead of relying on the propagation of the return value from `receiveRequest` to `sendResponse`, use the transaction object from the `Context` in `sendResponse` that you're already adding in `co.elastic.apm.agent.vertx_4.VertxWebHelper#startOrGetTransaction`.

As the `co.elastic.apm.agent.vertx_4.tracer.ElasticVertxTracer` doesn't activate the transaction, I suppose that you rely on `co.elastic.apm.agent.vertx_4.wrapper.RoutingContextHandlerWrapper` to activate the transaction when a route gets executed?

Reply:
Good point, resolved this as proposed.